File: scripts/MH_w_dnc.R
# This script implements the normal approximation approach and implements
# the divide and conquer algorithm

###################
#### Libraries ####
###################

library(dplyr)
library(bayestestR)
library(parallel)
library(coda)

###################
#### Functions ####
###################

log_post_fun <- function(param, x, y, ratio, sigma, mu) {
  (ratio)*sum(y*x%*%param - log(1 + exp(x%*%param))) -
    drop(0.5*t((param - mu)) %*% solve(sigma) %*% (param - mu))
}

#######################
#### Main function ####
#######################

inner_draws <- function(k_dat, NN = 1e4, total_n) {
  
  k_dat <- as.matrix(k_dat)
  
  # Set prior values
  mu <- rep(0, ncol(k_dat)) # need to add 1 for intercept and subtract one for y so it cancels out
  sd_temp <- sqrt(diag(var(k_dat[, -1])))
  scale <- sd_temp # scale continuous vars
  sigma <- diag(c(40^2, 3^2 * scale))  # flat priors on beta

  k_dat <- cbind(k_dat[,1], 1, k_dat[, 2:ncol(k_dat)])
  colnames(k_dat)[1] <- "y" 
  colnames(k_dat)[2] <- "intercept" 
  
  acc_count <- 0
  sd_prop <- 0.02
  glm_fit <- glm(y ~ -1 + ., as.data.frame(k_dat), family = "binomial")
  proposal_cov <- diag(diag(vcov(glm_fit)))*sd_prop
  
  beta_draws_mh <- matrix(0.0, NN, ncol(k_dat) -1)
  beta_draws_mh[1, ] <- coef(glm_fit)
  
  ################
  #### Run MH ####
  ################
  
  for(i in 2:NN) {
    
    beta_draws_mh[i,] <- beta_draws_mh[i - 1,]
    
    ## Draw beta
    beta_proposal <- MASS::mvrnorm(1, mu = beta_draws_mh[i,], Sigma = proposal_cov)
    
    acc_prob <- exp(log_post_fun(beta_proposal, 
                                 x = k_dat[, -1],
                                 y = k_dat[, 1], 
                                 ratio = total_n/nrow(k_dat), #n/m_j
                                 sigma = sigma,
                                 mu = mu) - 
                      log_post_fun(beta_draws_mh[i, ],
                                   x = k_dat[, -1],
                                   y = k_dat[, 1], 
                                   ratio = total_n/nrow(k_dat), #n/m_j
                                   sigma = sigma,
                                   mu = mu))
    
    if (runif(1) < acc_prob) {
      
      beta_draws_mh[i,] <- beta_proposal
      acc_count <- acc_count + 1
      
    }
    
  }
  
  return(list(beta_draws_mh  = beta_draws_mh,
              acc_prob = acc_count/(NN-1)))

}

############################
#### Parallel function #####
############################

par_fun <- function(K, NN = 1e4, burn_in_n = 1000){
  
  if(NN <= burn_in_n){
    stop("Discarding more than sampled")
  }
  
  ###################
  #### Load Data ####
  ###################
  
  mod_dat <- data.table::fread("/Users/atlan/Adv_computing/AdvComp_final_proj/data/sim_data.csv",
                               colClasses = "numeric",
                               showProgress = T) 
  
  mod_dat <- as.data.frame(mod_dat)
  total_n <- nrow(mod_dat)
  
  #########################################################
  #### Partition data into batches and set up clusters ####
  #########################################################
  
  split_dat <- split(mod_dat, rep(1:K, length.out = nrow(mod_dat), each = ceiling(nrow(mod_dat)/K)))
  
  cl <- makeCluster(min(detectCores(), K))
  clusterExport(
    cl,
    c("inner_draws", 
      "log_post_fun",
      "NN",
      "total_n"),
    envir = environment()
  )
  clusterEvalQ(cl, {
    library(MASS)
  })
  
  results_temp <- parLapply(
    cl,
    split_dat,
    function(x){inner_draws(k_dat = x, NN = NN, total_n = total_n)}
  )

  stopCluster(cl)
  gc()
  
  acc_probs <- sapply(1:length(results_temp), function(i) results_temp[[i]]$acc_prob)
  results <- lapply(1:length(results_temp), function(i) results_temp[[i]]$beta_draws_mh)
  
  ########################
  #### Remove burn-in ####
  ########################

  remove_burnin <- function(x, burnin) {
    x[-(1:burnin),]
  }
  
  results <- lapply(results, remove_burnin, burn_in_n)
  
  ########################
  #### Recenter Draws ####
  ########################
  
  subset_mean <- t(sapply(1:K, function(i) colMeans(results[[i]]))) # rows indicate subset
  global_mean <- colMeans(bind_rows(lapply(results, data.frame)))
  recenter <- t(sapply(1:K, function(i) subset_mean[i, ] - global_mean)) # rows indicate subset
  results_recentered <- lapply(1:K,function(i) results[[i]] - matrix(recenter[i, ],
                                                                     nrow = nrow(results[[i]]),
                                                                     ncol = ncol(results[[i]]),
                                                                     byrow = T))
  full_data_draws <- do.call(rbind, results_recentered)

  return(list(full_data_draws = full_data_draws,
              acc_probs = acc_probs))
}

set.seed(123123)

out <- par_fun(K=25)
full_data_draws <- out$full_data_draws
acc_probs <- out$acc_probs

gc()

time <- microbenchmark::microbenchmark(par_fun(K=25),
                                       unit = "s",
                                       times = 10)

elapsed <- tibble(summary(time)) %>% 
  dplyr::select(-expr, -neval) %>% 
  mutate(across(everything(), ~./60)) #convert to minutes

#################
#### Summary ####
#################

summary <- describe_posterior(as.data.frame(full_data_draws))
hpd <- hdi(as.data.frame(full_data_draws))

results <- tibble(var = summary[,1],
                  coef = summary[,2],
                  central_lower = summary[, 4],
                  central_upper= summary[, 5],
                  hpd_lower = hpd[,3],
                  hdp_upper = hpd[,4])

##############
#### Save ####
##############

out <- list(result_table = results,
            comp_time = elapsed,
            acc_probs = acc_probs,
            ESS = coda::effectiveSize(full_data_draws))

save(out, full_data_draws,
     file = "/Users/atlan/Adv_computing/AdvComp_final_proj/data/MH_dnc.Rdata")

File: scripts/MH_w_dnc.jl


using CSV, DataFrames, Statistics, Distributions, LinearAlgebra, Optim, NLSolversBase, BenchmarkTools, Random, Distributed, MLUtils, GLM

Threads.nthreads()

###################
#### Functions ####
###################

function log_post_fun(param, x, y, ratio, sigma, mu)
    p = 1 ./ (1 .+ exp.(-x * param))
    loglike = sum(logpdf.(Binomial.(1, p), y))
    logprior = logpdf(MvNormal(vec(mu), sigma), param)[1,1]
    return (ratio*loglike) + logprior
end
       
#######################
#### Main function ####
#######################

function inner_draws(k_dat, NN, total_n, burn_in)
  
  # Set prior values
  mu = zeros(size(k_dat)[2], 1)
  temp_sigma = Diagonal([40^2; 3^2 .* var.(eachcol(k_dat[:, 1:end .!= 1]))])  # flat priors on beta
  sigma = convert(Matrix,  temp_sigma)

  # Add vector of 1s for intercept 
  y = k_dat[:, 1]
  x = [ones(size(k_dat)[1]) k_dat[:, 2:size(k_dat)[2]]]
  ratio = total_n/size(k_dat)[1]

  ## Run MH
  acc_count = 0
  sd_prop = 0.02

  fm = @formula(x1 ~ x2 + x3 + x4+ x5)
  logit = glm(fm, DataFrame(k_dat), Binomial(), LogitLink())
  temp = GLM.coeftable(logit)
  proposal_cov = convert(Matrix, Diagonal(temp.cols[2].^2)*sd_prop)
 
  beta_draws_mh = zeros(NN, size(k_dat)[2])
  beta_draws_mh[1,:] = temp.cols[1]

  for i in 2:NN
    
    beta_draws_mh[i,:] = beta_draws_mh[i-1,:]

    ## Draw beta
    beta_proposal = rand(MvNormal(beta_draws_mh[i,:], proposal_cov), 1)

    acc_prob = exp(log_post_fun(beta_proposal, x, y, ratio, sigma, mu) - log_post_fun(beta_draws_mh[i,:], x, y, ratio, sigma, mu))
    
    if (rand(Uniform(0,1), 1)[1] < acc_prob)
      beta_draws_mh[i,:] = beta_proposal
      acc_count += 1
    end

  end  

  acc_prob = acc_count/NN

  #remove burn-in
  beta_draws_mh = beta_draws_mh[(burn_in+1):end, :]

  return beta_draws_mh, acc_prob

end

#######################
#### Main Function ####
#######################

function par_fun(NN = 10000, burn_in = 1000)
 
  if (burn_in > NN)
    return "Error: throwing out more than sampled"
  end

  ###################
  #### Load Data ####
  ###################
  
  mod_dat = CSV.read("/Users/atlan/Adv_computing/AdvComp_final_proj/data/sim_data.csv",
                     DataFrame, header = 1)

  total_n = size(mod_dat)[1]

  split_dat = chunk(Matrix(mod_dat), 25, dims=1)

  res = Array{Float64}(undef,NN-burn_in,size(mod_dat)[2],25)
  acc_count = Array{Float64}(undef,25,1)

  Threads.@threads for i in 1:25
    draws, acc_prob = inner_draws(split_dat[i], NN, total_n, burn_in)
    (@view acc_count[i]) .= acc_prob
    (@view res[:,:,i]) .= draws
  end
  
  ########################
  #### Recenter Draws ####
  ########################
  
  subset_mean = [mean(res[:,j,i]) for i in 1:size(res)[3], j in 1:size(res)[2]]
  weights =  [size(res[:,:, i])[1] for i in 1:size(res)[3]]
  weight_mat = reshape(weights, 1, :)/sum(weights)
  global_mean = vec(*(weight_mat, subset_mean))'
  
  for i in 1:size(res)[3]
    res[:,:,i] = res[:,:,i] .- subset_mean[i,:]' .+ global_mean
  end

  full_data_draws = reduce(vcat,(eachslice(res, dims=3)))

  return full_data_draws, acc_count

end

@time beta_draws, acc_prob  = par_fun(100, 10)
Random.seed!(123123)
time = @elapsed beta_draws, acc_prob = par_fun(10000, 1000)
time = round(time, digits = 4)
CSV.write("/Users/atlan/Adv_computing/AdvComp_final_proj/data/MH_dnc_$time.csv", 
          DataFrame(Matrix(beta_draws)))
CSV.write("/Users/atlan/Adv_computing/AdvComp_final_proj/data/MH_dnc_acc_prob_$time.csv", 
          DataFrame(Matrix(acc_prob)))

